pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,477 @@
+	if err != nil {
+		return err
+	}
+	_, _, normalizedValue, _, _ := splitUserInput(namedValue)
+	return sysregistriesv2.AddShortNameAlias(ctx, normalizedShortName, normalizedValue)
+}
+
+// Remove clears the short-name alias for the specified name.  It throws an
+// error in case name does not exist in the machine-generated
+// short-name-alias.conf.  In such case, the alias must be specified in one of
+// the registries.conf files, which is the users' responsibility.
+//
+// Note that tags and digests are stripped from the specified name.
+func Remove(ctx *types.SystemContext, name string) error {
+	isShort, shortRef, err := isShortName(name)
+	if err != nil {
+		return err
+	}
+	if !isShort {
+		return errors.Errorf("%q is not a short name", name)
+	}
+	_, _, normalizedShortName, _, _ := splitUserInput(shortRef)
+	return sysregistriesv2.RemoveShortNameAlias(ctx, normalizedShortName)
+}
+
+// Resolved encapsulates all data for a resolved image name.
+type Resolved struct {
+	PullCandidates []PullCandidate
+
+	userInput               string
+	unqualifiedSearchOrigin string
+	rationale               rationale
+	systemContext           *types.SystemContext
+}
+
+func (r *Resolved) addCandidate(named reference.Named) {
+	r.PullCandidates = append(r.PullCandidates, PullCandidate{named, r.userInput, false, r.systemContext})
+}
+
+func (r *Resolved) addCandidateToRecord(named reference.Named) {
+	r.PullCandidates = append(r.PullCandidates, PullCandidate{named, r.userInput, true, r.systemContext})
+}
+
+// Allows to reason over pull errors and add some context information.
+// Used in (*Resolved).WrapPullError.
+type rationale int
+
+const (
+	// No additional context.
+	rationaleNone rationale = iota
+	// Resolved value is a short-name alias.
+	rationaleAlias
+	// Resolved value has been completed with an Unqualified Search Registry.
+	rationaleUSR
+	// Resolved value has been selected by the user (via the prompt).
+	rationaleUserSelection
+)
+
+// Description returns a human-readable description about the resolution
+// process (e.g., short-name alias, unqualified-search registries, etc.).
+// It is meant to be printed before attempting to pull the pull candidates
+// to make the short-name resolution more transparent to user.
+//
+// If the returned string is empty, it is not meant to be printed.
+func (r *Resolved) Description() string {
+	switch r.rationale {
+	case rationaleAlias:
+		return fmt.Sprintf("Resolved short name %q to a recorded short-name alias (origin: %s)", r.userInput, r.unqualifiedSearchOrigin)
+	case rationaleUSR:
+		return fmt.Sprintf("Completed short name %q with unqualified-search registries (origin: %s)", r.userInput, r.unqualifiedSearchOrigin)
+	case rationaleUserSelection, rationaleNone:
+		fallthrough
+	default:
+		return ""
+	}
+}
+
+// FormatPullErrors is a convenience function to format errors that occured
+// while trying to pull one or more PullCandidates.
+func FormatPullErrors(pullErrors []error) error {
```

Comment:
This should be a method on `Resolved` so that it can benefit from all the private data in there, and the API should require `len(pullErrors) == len(Resolved.PullCandidates)`.